File: src/Day24.hs
{-|
Module:         Day24
Description:    <https://adventofcode.com/2019/day/24 Day 24: Planet of Discord>
-}
{-# LANGUAGE TupleSections #-}
module Day24 (day24a, day24b) where

import Common (findCycle, neighbors)
import Data.Ix (inRange, range)
import qualified Data.Map as Map (filterWithKey, fromListWith, keysSet)
import Data.Set (Set)
import qualified Data.Set as Set (fromDistinctAscList, mapMonotonic, member, size)

parse :: String -> Set (Int, Int)
parse input = Set.fromDistinctAscList $ do
    (y, line) <- zip [-2..2] $ lines input
    (x, '#') <- zip [-2..2] line
    return (y, x)

evolve :: (Ord a) => (a -> [a]) -> Set a -> Set a
evolve f s = Map.keysSet . Map.filterWithKey alive . Map.fromListWith (+) $
    (, 1 :: Int) <$> foldMap f s where
    alive _ 1 = True
    alive p 2 = not $ Set.member p s
    alive _ _ = False

withDepth :: Int -> (Int, Int) -> (Int, Int, Int)
withDepth z (x, y) = (x, y, z)

neighborsA :: (Int, Int) -> [(Int, Int)]
neighborsA = filter (inRange ((-2, -2), (2, 2))) . neighbors

neighborsB ::  (Int, Int, Int) -> [(Int, Int, Int)]
neighborsB (x, y, z) =
    map (withDepth z) (filter (/= (0, 0)) $ neighborsA (x, y)) ++
    [(x', 2 * y, z - 1) | x == 0, abs y == 1, x' <- [-2..2]] ++
    [(2 * x, y', z - 1) | abs x == 1, y == 0, y' <- [-2..2]] ++
    [(signum x, 0, z + 1) | abs x == 2] ++ [(0, signum y, z + 1) | abs y == 2]

day24a :: String -> Maybe Int
day24a = fmap count . findCycle . iterate (evolve neighborsA) . parse where
    count s = sum
      [ n
      | (p, n) <- zip (range ((-2, -2), (2, 2))) (iterate (* 2) 1)
      , Set.member p s
      ]

day24b :: String -> [Int]
day24b = fmap Set.size . iterate (evolve neighborsB) .
    Set.mapMonotonic (withDepth 0) . parse

File: src/aoc2019/day24.py
from collections import Counter
import fileinput


def parse(lines):
    return frozenset((x, y) for y in range(-2, 3) for x in range(-2, 3)
                     if lines[y + 2][x + 2] == '#')


def evolve(bugs, neighbors):
    counter = Counter()
    for bug in bugs:
        counter.update(neighbors(*bug))
    return frozenset(bug for bug, count in counter.items()
                     if count == 1 or count == 2 and bug not in bugs)


def part1(lines):
    '''
    >>> part1("....# #..#. #..## ..#.. #....".split())
    2129920
    '''
    def neighbors(x, y):
        return list(
            filter(None, ((x - 1, y) if x > -2 else None,
                          (x, y - 1) if y > -2 else None,
                          (x + 1, y) if x < 2 else None,
                          (x, y + 1) if y < 2 else None)))

    seen = set()
    bugs = parse(lines)
    while bugs not in seen:
        seen.add(bugs)
        bugs = evolve(bugs, neighbors)
    return sum(1 << x + 5 * y + 12 for y in range(-2, 3) for x in range(-2, 3)
               if (x, y) in bugs)


def part2(lines, n=200):
    '''
    >>> part2("....# #..#. #..## ..#.. #....".split(), 10)
    99
    '''
    def neighbors(x, y, z):
        neighbors = []
        if x > -2 and (x, y) != (1, 0): neighbors.append((x - 1, y, z))
        if y > -2 and (x, y) != (0, 1): neighbors.append((x, y - 1, z))
        if x < 2 and (x, y) != (-1, 0): neighbors.append((x + 1, y, z))
        if y < 2 and (x, y) != (0, -1): neighbors.append((x, y + 1, z))
        if (abs(x), y) == (1, 0):
            neighbors.extend((2 * x, y, z - 1) for y in range(-2, 3))
        if (x, abs(y)) == (0, 1):
            neighbors.extend((x, 2 * y, z - 1) for x in range(-2, 3))
        if abs(x) == 2: neighbors.append((x // 2, 0, z + 1))
        if abs(y) == 2: neighbors.append((0, y // 2, z + 1))
        return neighbors

    bugs = frozenset((x, y, 0) for x, y in parse(lines))
    for _ in range(n):
        bugs = evolve(bugs, neighbors)
    return len(bugs)


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))
